File: headers_list/src/main/java/CustomTypes/Point2D.java
package CustomTypes;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Point2D implements Comparable<Point2D> {

    private final double x;
    private final double y;

    private Point2D(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public static Point2D of(double x, double y) {
        return new Point2D(x, y);
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public double[] getCoords() {
        return new double[]{x, y};
    }

    public double getVector() {
        return Math.sqrt(x * x + y * y);
    }

    public static Point2D parse(String point) {
        if (point == null || point.trim().isEmpty()) {
            throw new IllegalArgumentException("Входная строка не может быть пустой.");
        }
        String[] parts = point.split(",");
        if (parts.length != 2) {
            throw new IllegalArgumentException("Строка должна быть в формате 'x,y'. Получено: " + point);
        }
        try {
            double x = Double.parseDouble(parts[0].trim());
            double y = Double.parseDouble(parts[1].trim());
            return Point2D.of(x, y);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Координаты должны быть числами. Получено: " + point, e);
        }
    }

    @Override
    public int compareTo(Point2D other) {
        return Double.compare(this.getVector(), other.getVector());
    }

    @Override
    public String toString() {
        return "Point2D(" + x + ", " + y + ")";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Point2D point2D = (Point2D) o;
        return Double.compare(point2D.x, x) == 0 && Double.compare(point2D.y, y) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }
}

File: headers_list/src/main/kotlin/CustomTypes/Point2D.kt
package CustomTypes

import java.util.Objects
import kotlin.math.sqrt

class Point2D private constructor(private val x: Double, private val y: Double) : Comparable<Point2D> {

    companion object {
        @JvmStatic
        fun of(x: Double, y: Double): Point2D {
            return Point2D(x, y)
        }

        @JvmStatic
        fun parse(point: String): Point2D {
            require(!point.trim().isEmpty()) { "Входная строка не может быть пустой." }
            
            val parts = point.split(",")
            require(parts.size == 2) { "Строка должна быть в формате 'x,y'. Получено: $point" }
            
            try {
                val x = parts[0].trim().toDouble()
                val y = parts[1].trim().toDouble()
                return of(x, y)
            } catch (e: NumberFormatException) {
                throw IllegalArgumentException("Координаты должны быть числами. Получено: $point", e)
            }
        }
    }

    fun getX(): Double {
        return x
    }

    fun getY(): Double {
        return y
    }

    fun getCoords(): DoubleArray {
        return doubleArrayOf(x, y)
    }

    fun getVector(): Double {
        return sqrt(x * x + y * y)
    }

    override fun compareTo(other: Point2D): Int {
        return this.getVector().compareTo(other.getVector())
    }

    override fun toString(): String {
        return "Point2D($x, $y)"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val point2D = other as Point2D
        return point2D.x.compareTo(x) == 0 && point2D.y.compareTo(y) == 0
    }

    override fun hashCode(): Int {
        return Objects.hash(x, y)
    }
}
